### PR TITLE
Fix iceworks-server unable to start if project directory is not present in local environment

### LINKED ISSUE #2585: [iceworks 3.0] iceworks-server unable to start if project directory is not present in local environment
## Steps to reproduce
1. Start `iceworks`. Make sure that there is at least 1 project in `iceworks`:
image

2. Delete the project from your local environment.

3. Stop running `iceworks-server` e.g. running `Ctrl + C`.

4. Start `iceworks-server` again by running `npm run dev`.

5. A similar error message would be shown and `iceworks-server` can't be started anymore:
image

## Suggestions
- Allow `iceworks-server` to still execute but indicate an error in the UI.

## Temporary workaround
- Delete your project entry in `~/.iceworks/db.json`.

### PR DESCRIPTION
**Pull request by tanhengyeow**
Fixes #2585

### CODE REVIEW

File: packages/iceworks-server/src/lib/plugin/project-manager/app.ts [hunk 1]
```diff
@@ -253,7 +257,14 @@ class ProjectManager extends EventEmitter {
 
   private async refresh(): Promise<Project[]> {
     return await Promise.all(
-      storage.get('projects').map(async (projectPath) => {
+      storage.get('projects').filter(projectPath => {
+        if (fs.existsSync(projectPath) && fs.existsSync(`${projectPath}/package.json`)) {
+          return true;
+        }
+        // If current project path or its package.json file does not exist, we delete the project from storage
+        this.deleteProject({ projectPath, deleteFiles: false });
```
Comment (alvinhui) [inline packages/iceworks-server/src/lib/plugin/project-manager/app.ts]: better to record in filter and delete after filter?

File: packages/iceworks-server/src/lib/plugin/project-manager/app.ts [hunk 2]
```diff
@@ -418,7 +433,7 @@ class ProjectManager extends EventEmitter {
    */
   public async deleteProject(params: { projectPath: string; deleteFiles?: boolean }): Promise<void> {
     const { projectPath, deleteFiles } = params;
-    this.projects = this.projects.filter(({ path }) => path !== projectPath);
+    this.projects = this.projects && this.projects.filter(({ path }) => path !== projectPath);
```
Comment (alvinhui) [inline packages/iceworks-server/src/lib/plugin/project-manager/app.ts]: Under what conditions would it be undefined?

File: packages/iceworks-server/src/lib/plugin/project-manager/app.ts [hunk 3]
```diff
@@ -428,6 +443,11 @@ class ProjectManager extends EventEmitter {
     const newPanelSettings = storage.get('panelSettings').filter(({projectPath: project}) => project !== projectPath);
     storage.set('panelSettings', newPanelSettings);
 
+    // if current project is not in local environment, delete it from storage
+    if (projectPath === storage.get('project')) {
```
Comment (alvinhui) [inline packages/iceworks-server/src/lib/plugin/project-manager/app.ts]: Please check 462 line of this file, looks like superfluous logic.

File: scripts/docConfig.js [hunk 4]
```diff
@@ -78,7 +78,7 @@ const docCategories = {
           'zh-cn': '进阶指南',
           'en-us': 'Advanced guide',
         },
-        dir: 'material',
+        dir: 'advanced',
```
[the remark this answers is not in the document]
Comment (chenbin92) [reply scripts/docConfig.js]: 他这个 MR 估计是基于 Master checkout 的

File: packages/iceworks-server/src/lib/plugin/project-manager/app.ts (hunk 2 above)
Comment (tanhengyeow) [reply packages/iceworks-server/src/lib/plugin/project-manager/app.ts]: When the currently selected project in `iceworks` is deleted from the local environment before `iceworks-server`is started, `this.projects` would be `undefined`. So we need to check it here.

File: scripts/docConfig.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (tanhengyeow) [reply scripts/docConfig.js]: Yup, will be rebasing my branch to track `iceworks/release-3.0.3`.

File: packages/iceworks-server/src/lib/plugin/project-manager/app.ts (hunk 2 above)
Comment (alvinhui) [reply packages/iceworks-server/src/lib/plugin/project-manager/app.ts]: `this.projects` define by `await this.refresh()`, I think should alway be a array.

Comment (alvinhui) [reply packages/iceworks-server/src/lib/plugin/project-manager/app.ts]: If  the currently selected project in iceworks is deleted from the local environment, you've already dealt with it at `getCurrent`.

File: packages/iceworks-server/src/lib/plugin/project-manager/app.ts [hunk 5]
```diff
@@ -251,9 +255,9 @@ class ProjectManager extends EventEmitter {
     this.app = app;
   }
 
-  private async refresh(): Promise<Project[]> {
+  private async refresh(projects: []): Promise<Project[]> {
```
Comment (alvinhui) [inline packages/iceworks-server/src/lib/plugin/project-manager/app.ts]: rename:  `createProjects` ? no idea...

File: packages/iceworks-server/src/lib/plugin/project-manager/app.ts [hunk 6]
```diff
@@ -262,8 +266,20 @@ class ProjectManager extends EventEmitter {
   }
 
   public async ready() {
+    const projects = storage.get('projects').map(projectPath => {
```
Comment (alvinhui) [inline packages/iceworks-server/src/lib/plugin/project-manager/app.ts]: It's better to encapsulate this part of the logic as `refresh`.

**Pull request merged**